File: forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2ResponseHandler.java
/*
 * Copyright (c) 2019 - 2022 ForgeRock. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package org.forgerock.android.auth;

import android.net.Uri;

import okhttp3.Response;

import org.forgerock.android.auth.exception.ApiException;
import org.forgerock.android.auth.exception.AuthenticationException;
import org.json.JSONObject;

import java.net.HttpURLConnection;
import java.util.Objects;

import static org.forgerock.android.auth.OAuth2.*;

/**
 * Implementation for handling {@link OAuth2Client} response, and provide feedback to the registered {@link FRListener}
 */
class OAuth2ResponseHandler implements ResponseHandler {

    private static final String TAG = OAuth2ResponseHandler.class.getSimpleName();

    /**
     * Handle Authorization response.
     *
     * @param response The response from /authorize endpoint
     * @param listener Listener for receiving OAuth APIs related changes
     */
    void handleAuthorizeResponse(Response response, FRListener<String> listener) {
        if (response.isRedirect()) {
            String location = response.header("Location");
            Uri redirect = Uri.parse(location);
            String code = redirect.getQueryParameter("code");
            if (code != null) {
                Listener.onSuccess(listener, code);
            } else {
                String errorDescription = redirect.getQueryParameter("error_description");
                Listener.onException(listener, new ApiException(response.code(), response.message(), errorDescription));
            }
            close(response);
        } else {
            handleError(response, listener);
        }
    }

    /**
     * Handle Token response
     *
     * @param response The response from /token endpoint
     */
    void handleTokenResponse(SSOToken sessionToken, Response response, String origRefreshToken, FRListener<AccessToken> listener) {
        if (response.isSuccessful()) {
            try {
                JSONObject jsonObject = new JSONObject(response.body().string());
                Logger.debug(TAG, "Access Token Received");
                Listener.onSuccess(listener, AccessToken.builder()
                        .idToken(jsonObject.optString(ID_TOKEN, null))
                        .value(jsonObject.getString(ACCESS_TOKEN))
                        .refreshToken(jsonObject.optString(REFRESH_TOKEN, origRefreshToken))
                        .scope(AccessToken.Scope.parse(jsonObject.optString(SCOPE, null)))
                        .tokenType(jsonObject.optString(TOKEN_TYPE, null))
                        .expiresIn(jsonObject.optLong(EXPIRES_IN, 0))
                        .sessionToken(sessionToken)
                        .build());
            } catch (Exception e) {
                Logger.debug(TAG, "Fail parsing returned Access Token: %s", e.getMessage());
                Listener.onException(listener, e);
            }
        } else {
            Logger.debug(TAG, "Exchange Access Token with Authorization Code failed.");
            handleError(response, listener);
        }
    }

    /**
     * Handle revoke token response
     *
     * @param response The response from the API
     * @param listener The Listener to listen for events
     */
    void handleRevokeResponse(Response response, FRListener<Void> listener) {
        if (response.isSuccessful()) {
            Logger.debug(TAG, "Revoke success");
            Listener.onSuccess(listener, null);
            close(response);
        } else {
            Logger.debug(TAG, "Revoke failed");
            handleError(response, listener);
        }
    }

}

File: forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2ResponseHandler.kt
/*
 * Copyright (c) 2019 - 2025 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package org.forgerock.android.auth

import android.net.Uri
import okhttp3.Response
import org.forgerock.android.auth.AccessToken.Scope.Companion.parse
import org.forgerock.android.auth.Logger.Companion.debug
import org.forgerock.android.auth.exception.ApiException
import org.json.JSONObject

/**
 * Implementation for handling [OAuth2Client] response, and provide feedback to the registered [FRListener]
 */
class OAuth2ResponseHandler : ResponseHandler {
    /**
     * Handle Authorization response.
     *
     * @param response The response from /authorize endpoint
     * @param state The state parameter included with the /authorize endpoint
     * @param listener Listener for receiving OAuth APIs related changes
     */
    fun handleAuthorizeResponse(response: Response,
                                state: String,
                                listener: FRListener<String>) {
        try {
            if (response.isRedirect) {
                val location = response.header("Location")
                val redirect = Uri.parse(location)
                val code = redirect.getQueryParameter("code")
                val responseState = redirect.getQueryParameter("state")
                if (responseState == null || responseState != state) {
                    Listener.onException(listener, IllegalStateException("OAuth2 state mismatch"))
                    return
                }
                if (code != null) {
                    listener.onSuccess(code)
                } else {
                    val errorDescription = redirect.getQueryParameter("error_description")
                    listener.onException(ApiException(
                        response.code, response.message, errorDescription))
                }
            } else {
                handleError(response, listener)
            }
        } finally {
            close(response)
        }
    }

    /**
     * Handle Token response
     *
     * @param response The response from /token endpoint
     */
    fun handleTokenResponse(sessionToken: SSOToken?,
                            response: Response,
                            origRefreshToken: String?,
                            listener: FRListener<AccessToken>) {
        if (response.isSuccessful) {
            try {
                response.body?.let { responseBody ->
                    val jsonObject = JSONObject(responseBody.string())
                    debug(TAG, "Access Token Received")
                    listener.onSuccess(AccessToken(
                        value = jsonObject.getString(OAuth2.ACCESS_TOKEN),
                        tokenType = jsonObject.optString(OAuth2.TOKEN_TYPE).takeIf { it.isNotEmpty() },
                        scope = parse(jsonObject.optString(OAuth2.SCOPE).takeIf { it.isNotEmpty() } ?: ""),
                        expiresIn = jsonObject.optLong(OAuth2.EXPIRES_IN, 0),
                        refreshToken = jsonObject.optString(OAuth2.REFRESH_TOKEN).takeIf { it.isNotEmpty() } ?: origRefreshToken,
                        idToken = jsonObject.optString(OAuth2.ID_TOKEN).takeIf { it.isNotEmpty() },
                        sessionToken = sessionToken))
                }
           } catch (e: Exception) {
                debug(TAG, "Fail parsing returned Access Token: %s", e.message)
                listener.onException(e)
            }
        } else {
            debug(TAG, "Exchange Access Token with Authorization Code failed.")
            handleError(response, listener)
        }
    }

    /**
     * Handle revoke token response
     *
     * @param response The response from the API
     * @param listener The Listener to listen for events
     */
    fun handleRevokeResponse(response: Response, listener: FRListener<Void?>) {
        if (response.isSuccessful) {
            debug(TAG, "Revoke success")
            listener.onSuccess(null)
            close(response)
        } else {
            debug(TAG, "Revoke failed")
            handleError(response, listener)
        }
    }

    companion object {
        private val TAG: String = OAuth2ResponseHandler::class.java.simpleName
    }
}

